### PR TITLE
Channel Listeners

### PR DESCRIPTION
**Pull request by Krzmbrzl**
Mumble_ChannelListeners

This implements #3319 by allowing users to "listen" to a channel they have not joined. Doing so will place a "listener proxy" (in other software this is sometimes known as a "phantom") in that channel which will look like a normal user except that it'll have the same name as the user listening to that channel and an ear-icon instead of the normal avatar-icon. It will also always show a muted-icon next to it.

If a listener proxy is in a channel, the server will route all audio packets from that channel to the user the proxy belongs to (as if that user was in the channel). Note though that the opposite of this is not true: The users in the channel will not hear audio from the listening user unless that user decides to join the channel.

Furthermore it is possible to set a local volume adjustment for each individual listener proxy so that audio received through them can be tuned in volume.

Finally the ACLEditor is changed so that every time you fiddle with the Enter privilege, you'll also set the Listen privilege automatically. For why I think this to be necessary see the commit message of 77e89e47568554187fa380868a1cb52aed73a257

This PR introduces 2 new server-settings that can be set in the `murmur.ini`:

| *Name* | Description |
| ---------- | --------------- |
| `listenersperchannel` | The amount of allowed listener proxies in a single channel. It defaults to `-1` meaning that there is no limit. Set to `0` to disable Channel Listeners altogether. |
| `listenersperuser` | The amount of listener proxies a single user may have. It defaults to `-1` meaning that there is no limit. Set to `0` to disable Channel Listeners altogether. |

---

ToDo:
- [x] Disable context-menu on listeners
- [x] Fix ear icon in modern theme
- [x] Add server config options to limit / disable channel listening
- [x] Add local volume factor for each virtual ear
- [x] Add dedicated listening ACL
- [x] Rework ACLEditor to go nicely with the new ACL
- [x] Use IDs instead of pointers for storing listening users or listener channels
- [x] Update translations

Changelog
```
| Added: ChannelListeners
```

### CODE REVIEW

**Comment by seantalts** [comment]
I have a dumb question - when I try to build this I get `RCC: Error in '../../themes/MumbleTheme.qrc': Cannot find file 'Mumble/ear.svg'`, which looks like it's actually in ./themes/Classic/ear.svg. Yet it passed all of CI on Github. Any tips for what I'm doing wrong?

**Comment by Krzmbrzl** [comment]
@seantalts you probably haven't initialized the `themes/Mumble` submodule. That's where the more modern (and default) theme is located.
We are currently maintaining 2 separate themes. The classic (old) one in `themes/Classic` and the newer one which actually has its own repo at https://github.com/mumble-voip/mumble-theme

**Comment by seantalts** [comment]
Thanks, looks like `git submodule update --init --recursive` fixed it for me; I think I forgot to do that step listed here https://wiki.mumble.info/wiki/BuildingLinux#Getting_the_source

**Comment by wfjsw** [comment]
It seems this change is not backward compatible, that means in no way can I know who is listening with a older version of Mumble.

**Comment by Krzmbrzl** [comment]
@wfjsw yes this is correct. However there is no way one could have made this backwards compatible. It is now up to the server administrator. If (s)he knows that a lot of people are using a client version lower than 1.4.0, then the admin can disable the ChannelListener feature via the provided config options.

**Comment by streaps** [comment]
Would it possible that a listener proxy can only be created in channels with 100% 1.4 clients? Or maybe the server could send a notification message for older clients?

This is a very useful feature, but it's little bit spooky / nontransparent, if users don't know that a phantom is listening.

**Comment by Krzmbrzl** [comment]
> Would it possible that a listener proxy can only be created in channels with 100% 1.4 clients?

This would probably render this feature useless for several years on a server with a decent amount of traffic.

> Or maybe the server could send a notification message for older clients?

I guess that could work. Sending a TextMessage from the server to all clients in an affected channel should be doable :thinking:

**Comment by Krzmbrzl** [comment]
Though actually when giving it some more thought this would require some ugly extra-code in the server code... Not really sure if this is worth it

**Comment by streaps** [comment]
> > Would it possible that a listener proxy can only be created in channels with 100% 1.4 clients?
> 
> This would probably render this feature useless for several years on a server with a decent amount of traffic.

It could be enabled/disabled by a config option. But I think I'm starting to dislike the idea anyway. Now it works - now it doesn't work - now it works again... looks more a bug like a feature from the user's perspective.

**Comment by streaps** [comment]
> Though actually when giving it some more thought this would require some ugly extra-code in the server code... Not really sure if this is worth it

I guess not.

I don't know when 1.4.0 is going to be released (in 10 years or earlier ;). If it's not in the next couple of months, maybe a backport of the functionality that only show the listener proxy in the channel (👂 username) to the 1.3.x branch would be less ugly. So people who use 1.3.something could already see the listener proxy (but have to upgrade to 1.4.x to create one or set the ACLs).

It's just an idea, which might be impractical or also not worth the effort.

**Comment by Krzmbrzl** [comment]
> in 10 years or earlier

Hopefully sooner :P

> maybe a backport of the functionality that only show the listener proxy in the channel

Well I guess that'd certainly solve the problem but it would not fit our strategy of backporting at all. We only backport bug and security fixes but no features...


What could be easily done would be to check whether the server has CHannelListeners enabled and if it does, it can send a text-message to a client that is connecting with a Mumble version less than 1.4.0 to inform them that the new feature is enabled on this server but that they won't see it in their client version. Would that be enough?

**Comment by streaps** [comment]
Sounds like a good compromise and an incentive for users to upgrade to a newer Mumble client version.

**Comment by jeffmikels** [comment]
What's the protobuf command for listening in on a new channel?

**Comment by Krzmbrzl** [comment]
What protobuf command? There are no protobuf commands :thinking:

Are you referring to the gRPC interface? If so: it's not implemented

**Comment by absfilho2** [comment]
Hello,

I am also from the boardcast world and this function of just listening to other channels is very interesting.
Could you put a shortcut so you can start and end listening to a certain channel?

Another thing, when do you want to make the version with this function a new reliased one?

**Comment by Krzmbrzl** [comment]
> Could you put a shortcut so you can start and end listening to a certain channel?

That could certainly be done, yes. This should probably be its own issue here on GitHub so we don't lose track of this feature request :)

> Another thing, when do you want to make the version with this function a new reliased one?

We don't have an ETA. We wanted to start distributing snapshot releases a few weeks ago but we are currently fighting some issues with the Windows installer. Once that's resolved, there should be snapshots available

**Comment by davidebeatrici** [comment]
For reference: #4488

File: src/murmur/Messages.cpp [hunk 1]
```diff
@@ -848,7 +888,33 @@ void Server::msgUserState(ServerUser *uSource, MumbleProto::UserState &msg) {
 		userEnterChannel(pDstServerUser, c, msg);
 		log(uSource, QString("Moved %1 to %2").arg(QString(*pDstServerUser), QString(*c)));
 		bBroadcast = true;
+
+		if (ChannelListener::isListening(pDstServerUser, c)) {
+			// If a user joins a channel (s)he has been listening to before, it means that this user will no
```
Comment (kripton) [inline src/murmur/Messages.cpp:893]: Has this been tested? At least in the UI this is not reflected. I've tested it locally (be in ChanA, listen to ChanB and then join ChanB). Result:
grafik

Comment (Krzmbrzl) [reply src/murmur/Messages.cpp:893]: This has been changed later on. The restriction of not being allowed to listen to the channel you're currently in, has been lifted in #4129

Comment (kripton) [reply src/murmur/Messages.cpp:893]: Oh, great :) True, I didn't check the current state of the code but the explanation in #4129 makes sense. Thanks for explaining :+1:

Comment (Krzmbrzl) [reply src/murmur/Messages.cpp:893]: Sure, no problem :)

**Comment by SNRSE** [comment]
Do i understand correctly that this should work now in the latest version ?

**Comment by Krzmbrzl** [comment]
It works in the latest snapshot (1.4.0) but only on servers that are running at least 1.4.0 as well.

**Comment by Th3irdEye** [comment]
Any way this could be set up so the listener icon and user don't appear in a channel until at least 1 real user is in the channel?

**Comment by Krzmbrzl** [comment]
> Any way this could be set up so the listener icon and user don't appear in a channel until at least 1 real user is in the channel?

Depends on your definition of "set up". If that involves making changes to the Mumble source code, then yeah that is probably possible but if you are looking for a way to achieve this using only settings in the client, I don't think this is possible :thinking:

**Comment by Th3irdEye** [comment]
Thanks. I was talking about either client or server side settings as they are now. Maybe a good an idea for a future update then. I have a listener set up on my recording PC that monitors a couple of recording specific channels and it would be much cleaner for the server layout to not show the listener until necessary imo. I am happy I don't have to link channels to get similar functionality anymore though.

**Pull request merged**